ext/session/session.c

```diff
@@ -2344,15 +2782,64 @@ static PHP_FUNCTION(session_start)
 }
 /* }}} */
 
-/* {{{ proto bool session_destroy(void)
+
+/* {{{ proto array session_info(void)
+   Return session internal data information */
+static PHP_FUNCTION(session_info)
+{
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	if (PS(session_status) != php_session_active) {
+		php_error_docref(NULL, E_WARNING, "Session is not active");
+		RETURN_FALSE;
+	}
+
+	ZEND_ASSERT(!Z_ISUNDEF(PS(internal_data)));
+	ZVAL_COPY(return_value, &PS(internal_data));
+	return;
+}
+/* }}} */
+
+
+/* {{{ proto int session_gc(void)
+   Perform session GC manually */
+static PHP_FUNCTION(session_gc)
+{
+	zend_long nrdels;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	if (PS(session_status) != php_session_active) {
+		php_error_docref(NULL, E_WARNING, "Session is not active");
+		RETURN_FALSE;
+	}
+
+	nrdels = php_session_gc();
+	if (nrdels < 0) {
+		RETURN_FALSE;
+	}
+	ZVAL_LONG(return_value, nrdels);
+}
+/* {{{ proto bool session_destroy([bool immediate])
    Destroy the current session and all data associated with it */
 static PHP_FUNCTION(session_destroy)
 {
-	if (zend_parse_parameters_none() == FAILURE) {
+	zend_bool immediate=0;
```

Comment:
How about change this patch behaves like currently?

We have session.use_strict_mode=0 as the default now. Therefore, if a script destroy the session, another scripts executed concurrently for the session will get empty session with the same session ID. Therefore, client(browser) will not flood of Set-Cookie headers and Cookie header won't be empty string. This is not optimal behavior, though. (session.use_strict_mode=1 does not allow user supplied uninitialized session ID and generates new session ID. Set-Cookie may sent multiple times in a short period. This causes Cookie header having null value.)

I'll change the code so that reset $_SESSION to have empty array, then we can make sure destroyed session never leaks sensitive data/state.